services/user-authenticator.js

```diff
@@ -0,0 +1,91 @@
+'use strict';
+/* jshint sub: true */
+var _ = require('lodash');
+var bcrypt = require('bcryptjs');
+var jwt = require('jsonwebtoken');
+var AllowedUsersFinder = require('../services/allowed-users-finder');
+var RefreshTokenSender = require('../services/refresh-token-sender');
+var EnvironmentExpirationTime = require('../services/environment-expiration-time');
+var uuidV1 = require('uuid/v1');
+
+function UserAuthenticator(request, opts) {
+  this.perform = function () {
+    return new AllowedUsersFinder(request.body.renderingId, opts)
+      .perform()
+      .then(function (allowedUsers) {
+        if (!opts.authSecret) {
+          throw new Error('Your Forest authSecret seems to be missing. Can ' +
+            'you check that you properly set a Forest authSecret in the ' +
+            'Forest initializer?');
+        }
+
+        if (allowedUsers.length === 0) {
+          throw new Error('Forest cannot retrieve any users for the project ' +
+            'you\'re trying to unlock.');
+        }
+
+        var user = _.find(allowedUsers, function (allowedUser) {
+          return allowedUser.email === request.body.email;
+        });
+
+        if (user === undefined) { throw new Error(); }
+
+        if (request.body.refreshToken) { return user; }
```

Comment:
You need to make this `if` condition very restrictive to prevent hackers without password (but with a random `refreshToken`) to get an access token.